core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -0,0 +1,481 @@
+     */
+    public static Options defaultOptions() {
+      return new Options();
+    }
+  }
+
+  private final ReentrantLock mLock = new ReentrantLock();
+  private final Condition mNotEmpty = mLock.newCondition();
+  private final int mMaxCapacity;
+  private final int mMinCapacity;
+
+  // Tracks the resources that are available ordered by lastAccessTime (the first one is
+  // the most recently used resource).
+  @GuardedBy("mLock")
+  private TreeSet<ResourceInternal<T>> mResourceAvailable =
+      new TreeSet<>(new Comparator<ResourceInternal<T>>() {
+        @Override
+        public int compare(ResourceInternal<T> c1, ResourceInternal<T> c2) {
+          if (c1 == c2) {
+            return 0;
+          }
+          if (c1.mLastAccessTimeMs == c2.mLastAccessTimeMs) {
+            return c1.mIdentity - c2.mIdentity;
+          }
+          return (int) (c2.mLastAccessTimeMs - c1.mLastAccessTimeMs);
+        }
+      });
+
+  // Tracks all the resources that are not closed.
+  @GuardedBy("mLock")
+  private HashMap<T, ResourceInternal<T>> mResources = new HashMap<>(32);
+
+  // Thread to scan mResourceAvailable to close those resources that are old.
+  private ScheduledExecutorService mExecutor;
+
+  /**
+   * Creates a dynamic pool instance.
+   *
+   * @param options the options
+   */
+  public DynamicResourcePool(Options options) {
+    mExecutor = new ScheduledThreadPoolExecutor(1);
+
+    mMaxCapacity = options.getMaxCapacity();
+    mMinCapacity = options.getMinCapacity();
+
+    mExecutor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        List<T> resourcesToGc = new ArrayList<T>();
+
+        try {
+          mLock.lock();
+          if (mResources.size() <= mMinCapacity) {
+            return;
+          }
+          Iterator<ResourceInternal<T>> iterator = mResourceAvailable.iterator();
+          while (iterator.hasNext()) {
+            ResourceInternal<T> next = iterator.next();
+            if (shouldGc(next)) {
+              resourcesToGc.add(next.mResource);
+              iterator.remove();
+              mResources.remove(next.mResource);
+            }
+          }
+        } finally {
+          mLock.unlock();
+        }
+
+        for (T resource : resourcesToGc) {
+          LOG.info("Resource {} is garbage collected.", resource.toString());
+          closeResource(resource);
+        }
+      }
+    }, options.getInitialDelayMs(), options.getGcIntervalMs(), TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Acquire a resource of type {code T} from the pool.
+   *
+   * @return the acquired resource
+   * @throws IOException if it fails to acquire because of the failure to create a new resource
+   */
+  @Override
+  public T acquire() throws IOException {
+    try {
+      return acquire(100  /* no timeout */, TimeUnit.DAYS);
+    } catch (TimeoutException e) {
+      LOG.error("Never should timeout in acquire().");
+      throw new RuntimeException(e);
```

Comment:
Is Throwables.propagate deprecated (Jiri probably)?  I was once told not to use it. But i cannot find reference for it. 

So i just changed it what you suggested.